### DISCUSSION TITLE
Wait in the background for response from the server and wait for an event to occur

### DISCUSSION THREAD

**Discussion by marek22k** [socketry/async#283] [Q&A]
Hello,

I would like to implement a client app. This can get one or more messages from the server, which should be evaluated in the background, so that the user can see the results later.

Furthermore I would like to use the Async Gem for this. What would be the correct usage here to use the Async Gem? Would I just write `Async` instead of `DoInBackground`?

Furthermore, there may be places in the program where I want to wait for a response to come in the server. Is there (maybe also with Async?) a more elegant method than running a loop until it occurs?

```ruby
class MyClient

  attr_reader :server_version

  def initialize
    @client = TCPSocket.new('localhost', 8888)
  end
  
  def run
    Background do
      while @client.data_available?
        if @client.gets.chomp == 'RequestComplete'
          @server_version = @client.gets.chomp
        end
      end
    end
  end
  
  def send_request
    @client.puts 'NewRequest'
  end
  
  def wait_until_version_available
    while server_version == nil
    
    end
  end

end

Async do
  client = MyClient.new
  client.run
  # do stuff
  client.send_request
  # do stuff
  client.wait_until_version_available
end
```

I am fairly new to programming with async and non-imperative programs. I hope this is the right place to ask.

**Comment by ioquatix** [comment]
Yes that looks okay to me, in principle.
